gc.c

```diff
@@ -1812,53 +1811,46 @@ static const rb_data_type_t id_to_obj_tbl_type = {
 };
 
 static VALUE
-object_id(VALUE obj)
+object_id0(VALUE obj)
 {
     VALUE id = Qfalse;
     rb_shape_t *shape = rb_obj_shape(obj);
-    unsigned int lock_lev;
 
-    // We could avoid locking if the object isn't shareable
-    // but we'll lock anyway to lookup the next shape, and
-    // we'd at least need to generate the object_id using atomics.
-    lock_lev = rb_gc_vm_lock();
+    if (rb_shape_has_object_id(shape)) {
+        rb_shape_t *object_id_shape = rb_shape_object_id_shape(obj);
+         id = rb_obj_field_get(obj, object_id_shape);
+         RUBY_ASSERT(id, "object_id missing");
+         return id;
+    }
 
-    if (rb_shape_too_complex_p(shape)) {
-        st_table *table = ROBJECT_FIELDS_HASH(obj);
-        if (rb_shape_has_object_id(shape)) {
-            st_lookup(table, (st_data_t)ruby_internal_object_id, (st_data_t *)&id);
-            RUBY_ASSERT(id, "object_id missing");
+    // TODO: can we do 64bit atomics? Or accept that `object_id_counter`
+    // is sometimes only 32 bits?
+    unsigned int lock_lev = rb_gc_vm_lock();
+    id = ULL2NUM(++object_id_counter * OBJ_ID_INCREMENT);
+    rb_gc_vm_unlock(lock_lev);
```

Comment:
@jhawthorn no rush, but I wonder if you have opinions about what is best to do here.

`rb_atomic_t` is 32bits on i686 but also on Windows AFAICT.

Reply:
Actually, I suppose I can just check for the size of `rb_atomic_t` and go the atomic route if it's large enough.